bin/transform-genbank

```diff
@@ -0,0 +1,317 @@
+        # Both location and division are not available
+        if pd.isna(value[1]):
+            location = division = None
+        # Only division is available
+        elif ',' not in value[1]:
+            location = None
+            division = value[1]
+        # Location and division are both available
+        elif ',' in value[1]:
+            location,division = value[1].split(',', 1)
+        # Unknown format of location data
+        else:
+            assert False, f"Found unknown format for geographic data: {value}"
+
+        # Convert US state codes to full names
+        if value[0] == 'USA' and division and us_states.get(division.strip()):
+            division = us_states[division.strip()]
+
+        location = location.strip().lower().title() if location else None
+        division = division.strip().lower().title() if division else None
+
+        divisions.append(division)
+        locations.append(location)
+
+    genbank_data['country']     = geographic_data[0]
+    genbank_data['division']    = pd.Series(divisions)
+    genbank_data['location']    = pd.Series(locations)
+
+    return genbank_data
+
+
+def parse_authors(genbank_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Abbreviate the column named `authors` to be "<first author> etl al" rather than a
+    full list. Returns the modified DataFrame.
+    """
+    # Strip and normalize whitespace
+    genbank_data['authors'] = genbank_data['authors'].str.replace(r'\s+', ' ')
+    # Strip to string before first comma
+    genbank_data['authors'] = genbank_data['authors'].str.replace(r"^([^,]+),.+", lambda m: m.group(1))
+    # Add et al to authors if authors is not an empty string
+    genbank_data.loc[genbank_data['authors'] != '', 'authors'] = genbank_data['authors'].astype(str) + ' et al'
+    # Replace blank authors with '?'
+    genbank_data.loc[genbank_data['authors'] == '', 'authors'] = '?'
+
+    return genbank_data
+
+
+def format_date_columns(genbank_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Format the date columns of *genbank_data* to have the format
+    %Y-%m-%d and return the modified DataFrame.
+    """
+    date_columns = ['date', 'date_submitted']
+
+    for column in date_columns:
+        genbank_data[column] = pd.to_datetime(genbank_data[column]).dt.strftime('%Y-%m-%d')
+
+    return genbank_data
+
+
+def generate_hardcoded_metadata(genbank_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Returns a DataFrame with a column for GenBank accession plus
+    additional columns containing hardcoded metadata.
+    """
+    hardcoded_metadata = pd.DataFrame(genbank_data['genbank_accession'])
+    hardcoded_metadata['virus']             = 'ncov'
+    hardcoded_metadata['gisaid_epi_isl']    = '?'
+    hardcoded_metadata['segment']           = 'genome'
+    hardcoded_metadata['age']               = '?'
+    hardcoded_metadata['sex']               = '?'
+    hardcoded_metadata['originating_lab']   = '?'
```

Comment:
age, sex, and originating lab seems to be available for some records in BioSample. However, the `Entrez.efetch` API does not work with the BioSample accession that we get from the GenBank pull. The only way I've been able to get this to work is to use `Entrez.esearch` to search the BioSample accession to get the BioSample ID, then use that BioSample ID to get the metadata via `efetch`. The caveat here is that I can't seem to use `esearch` to do a search for a batch of accessions, so we would need to make a separate `esearch` request for each accession, then do a batch `efetch` with all the returned BioSample IDs.

Reply:
Another discovery, `Entrez.read()` does not work with BioSample's XML (see: https://github.com/biopython/biopython/issues/2817). So we would have use a python XML parser to parse the data returned by `efetch`.

Reply:
I believe you can use `esearch` to find a batch of accessions by constructing a query string like this:

    SAMPLE1[accn] or SAMPLE2[accn] or …

The raw link for example: <https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=biosample&term=SAMEA2467098[accn]+or+SAMN07663113[accn]&retmode=json>

A long list of accessions will hit URL-length limits, so best to use POST params instead of GET + query params. I don't know if `Entrez.esearch` uses GET or POST by default or if it's dynamic depending on the data. It's also possible to use the [server-side search history feature](https://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2.Using_the_Entrez_History_Server) of Entrez to store the results of your esearch there and then tell efetch to use that saved id list, if you want to avoid round-tripping them.

Reply:
Thanks for digging into this! Looking at the [source code](https://biopython.org/DIST/docs/api/Bio.Entrez-pysrc.html) it seems that it dynamically uses POST if the query is over 1000 characters.

Reply:
Ahh, you're correct. I missed that!